internal/providers/zvm/zvm.go

```diff
@@ -0,0 +1,130 @@
+// Copyright 2016 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// The QEMU provider fetches a local configuration from the firmware config
+// interface (opt/com.coreos/config).
+
+package zvm
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"strings"
+
+	"github.com/coreos/ignition/v2/config/shared/errors"
+	"github.com/coreos/ignition/v2/config/v3_1_experimental/types"
+	"github.com/coreos/ignition/v2/internal/log"
+	"github.com/coreos/ignition/v2/internal/providers/util"
+	"github.com/coreos/ignition/v2/internal/resource"
+	"github.com/coreos/vcontext/report"
+)
+
+const authorizedSenders string = "*"
+
+func FetchConfig(f *resource.Fetcher) (types.Config, report.Report, error) {
+	/*fetch config files directly from reader device.*/
+	_, err := f.Logger.LogCmd(exec.Command("/sbin/modprobe", "vmur"), "loading zVM control program module")
+	if err != nil {
+		f.Logger.Err("couldn't install vmur module: %v", err)
+		return types.Config{}, report.Report{}, err
+	}
+	/* onlineDevice 000c */
+	logger := f.Logger
+	rc := onlineDevice(logger)
+	if !rc {
+		return types.Config{}, report.Report{}, err
+	}
+	f.Logger.Info("The reader device has been online.")
+	/* pullReader */
+	var comm string = "/sbin/vmur li | tail -n +2 | cut -c 1-80 | awk '{print $1,$2,$3,$10,$11}'"
+	cmd := exec.Command("/bin/sh", "-c", comm)
+	out, err := cmd.Output()
+	if err != nil {
+		f.Logger.Err("Can not get reader device: %v", err)
+	}
+	f.Logger.Info("The reader device info: ")
+	fmt.Println(string(out))
+	items := strings.Split(string(out), "\n")
+	/* Process each spool file that is class "x".*/
+	for _, records := range items {
+		record := strings.Split(records, " ")
+		f.Logger.Info("The single record info: ")
+		fmt.Println(record, len(record))
+		if len(record) == 1 {
```

Comment:
This length-check verify that it is the last null record to avoid the “record[1]” error.

Reply:
I don't understand? What is the `record[*]` error? We should still check that `len(record) == 5` (or is it `>=5`? I don't know the format) before accessing `record[4]` to eliminate the chance of panics from out of bounds access.

Reply:
Can you clarify about my previous question?

Reply:
The length of the record is >=11, but when I use strings.Split(s, "\n"). There will be a null record as the last record. It will cause error when read record[1].